src/Database/LSMTree/Internal/Lookup.hs

```diff
@@ -30,3 +60,171 @@ prepLookup :: Run fd -> SerialisedKey -> Maybe PageSpan
+  hashes, which means that there is little to no spatial locality when
+  performing lookups, and so we can skip looking at the page cache.
+-}
+
+newtype BatchSize = BatchSize { unBatchSize :: Int }
+
+-- | A batch lookup read fewer bytes than requested.
+data NotEnoughBytes = NotEnoughBytes
+  deriving (Show, Exception)
+
+{-# SPECIALIZE lookupsInBatches ::
+       HasBlockIO IO HandleIO
+    -> Resolve1
+    -> BatchSize
+    -> [Run (Handle HandleIO)]
+    -> [SerialisedKey]
+    -> IO [(SerialisedKey, Maybe (Entry SerialisedValue BlobSpan))]
+  #-}
+-- | Batched lookups.
+--
+-- When the length of the list of lookup keys exceeds the batch size, the list
+-- is subdivided into batches of at most the given batch size. Each batch is
+-- submitted concurrently using async IO, which could result in a speedup if
+-- this code is run with multi-threading enabled.
+--
+-- See Note [Batched lookups, buffer strategy and restrictions]
+--
+-- TODO: optimise by reducing allocations, possibly looking at core, or
+-- revisiting the batching approach.
+lookupsInBatches ::
+     forall m h. (MonadAsync m, PrimMonad m, MonadThrow m)
+  => HasBlockIO m h
+  -> Resolve1
+  -> BatchSize
+  -> [Run (Handle h)]
+  -> [SerialisedKey]
+  -> m [(SerialisedKey, Maybe (Entry SerialisedValue BlobSpan))]
+lookupsInBatches hbio resolve1 n runs ks = do
+    let prepped = prepLookups runs ks
+        (fioops, npages) = prepIOOpReads (Compose prepped)
+    mbuf <- newPinnedByteArray (npages * 4096)
+    let Compose ioops = fioops mbuf
+    let batches = groupsOfN (unBatchSize n) ioops
+    kess <- forConcurrently batches $ \batch -> do
+      ress <- submitIO hbio (fmap snd batch)
+      buf <- unsafeFreezeByteArray mbuf
+      forM (zip batch ress) $ \((k, ioop), res) -> do
+        unless (checkIOResult ioop res) $ throwIO NotEnoughBytes
+        let rpl = rawPageLookup (makeRawPage buf (unBufferOffset $ ioopBufferOffset ioop)) k
+        pure . (k,) $ case rpl of
+          LookupEntryNotPresent   -> Nothing
+          LookupEntry e           ->
+            -- Laziness ensures that we only compute the forcing of the value in
+            -- the entry when the result is needed.
+            Just (forceEntry e)
+          LookupEntryOverflow e m ->
+            let vOverflow = RawBytes (P.Vector
+                                        (unBufferOffset (ioopBufferOffset ioop) + 4096)
+                                        (fromIntegral m - 4096)
+                                        buf)
+            in  -- Laziness ensures that we only compute the appending of the
+                -- prefix and suffix when the result is needed. We do not use
+                -- 'force' here, since appending already creates a new primary
+                -- vector. TODO: verify if appending always computes a new
+                -- primary vector
+                Just $ first (\(SerialisedValue v) -> SerialisedValue (v <> vOverflow)) e
```

Comment:
The two parts are already adjacent in the buffer, so I wonder if `LookupEntryOverflow` should just return a value offset, so we can construct (and force) a single vector. This would also avoid the potential footgun of it currently returning a `SerialisedValue` that is in fact only a prefix (so it can't be used in the same way as others). :thinking: